sympy/combinatorics/pc_groups.py

```diff
@@ -0,0 +1,494 @@
+        (2, 9)
+
+        """
+        low = -1
+        high = -1
+        for i in range(len(word)-len(w)+1):
+            if word.subword(i, i+len(w)) == w:
+                low = i
+                high = i+len(w)
+                break
+        if low == high == -1:
+            return -1, -1
+        return low, high
+
+    def map_relation(self, w):
+        """
+        Return a conjugate relation.
+        Given a word formed by two free group elements, the
+        corresponding conjugate relation with those free
+        group elements is formed and mapped with the collected
+        word in the polycyclic presentation.
+
+        Examples
+        ========
+        >>> from sympy.combinatorics.pc_groups import Collector
+        >>> from sympy.combinatorics.free_groups import free_group
+        >>> F, x0, x1, x2 = free_group("x0, x1, x2")
+        >>> pc_relators = {x0**-1*x1*x0: x1**2, x1**-1*x2*x1: x2, x0**-1*x2*x0: x2*x1}
+        >>> relative_order = [2, 2, 3]
+        >>> word = x2**2*x1**7
+        >>> group = word.group
+        >>> collector = Collector(pc_relators, relative_order, group)
+        >>> w = x2*x1
+        >>> collector.map_relation(w)
+        x2
+        >>> w = x1*x0
+        >>> collector.map_relation(w)
+        x1**2
+
+        """
+        group = w.group
+        gens = sorted(w.contains_generators())
+        key = gens[0]**-1*gens[1]*gens[0]
+        return self.pc_relators[key]
+
+
+    def collected_word(self, word):
+        """
+        Return the collected form of a word.
+
+        A word `w` is called collected, if `w = x{i_1}**a_1*...*x{i_r}**a_r`
+        with `i_1 < i_2< ... < i_r` and `a_j` is in `{1, ..., s_j-1}`
+        if `s_j != infinity`.
+        Otherwise w is uncollected.
+
+        Examples
+        ========
+        >>> from sympy.combinatorics.pc_groups import Collector
+        >>> from sympy.combinatorics.free_groups import free_group
+        >>> F, x1, x2 = free_group("x1, x2")
+        >>> pc_relators = {x1**2 : (), x1*x2*x1**-1 : x2**-1, x1**-1*x2*x1 : x2**-1}
+        >>> relative_order = [2, None]
+        >>> word = x2**2*x1**7
+        >>> group = word.group
+        >>> collector = Collector(pc_relators, relative_order, group)
+        >>> collector.collected_word(word)
+        x1*x2**-2
+
+        """
+        group = self.group
+        while True:
+            w = self.minimal_uncollected_subword(word)
+            if not w:
+                break
+
+            low, high = self.subword_index(word, group.dtype(w))
+            if low == -1:
+                continue
+            s, e = w[0]
+            if len(w) == 2 and w[1][1] > 0:
+                gens = list(sorted(group.dtype(w).contains_generators()))
+                word_ = self.map_relation(group.dtype(w))
+                word_ = gens[0]*word_**e
+                word_ = group.dtype(word_)
+                word = word.substituted_word(low, high, word_)
+
+            elif len(w) == 2 and w[1][1] < 0:
```

Comment:
This part will not be needed for groups with finite relative orders as negative exponents can always be made positive. Therefore the minimal uncollected words of length 1 should be handled first. (Besides, for this part we will also need the alternate conjugate relations for `xj*xi*xj**-1` that are not necessary for finite groups.)

Reply:
Should we try removing this `elif`part?

Reply:
I am not sure. Perhaps it could be left there but it cannot be used before the alternate conjugate relations have been constructed. For finite groups, we should always first remove the negative exponents so the `elif` part would not be run.

Reply:
Are alternate conjugate relations needed only for infinite groups? then, we may not need to compute it because polycyclic groups are constructed only for finite groups. We may try to swap this `elif` with the last `if`.

Reply:
They are not needed for finite groups because the relative orders are finite and therefore all negative exponents can be collected. That could probably be the first task.